<launch>
<!-- Arguments Start -->

  <!-- Simulation -->
  <arg name="runsim"                default="true"/>
  <arg name="log"                   default="true"/>

  <!-- <arg name="mission"               default="mission_8agents_15.json"/> -->
  <arg name="mission"               default="mission_1agents_MPF.json"/>
  <arg name="mission_follwer"       default="mission_1agents_MPF2.json"/>
  <arg name="replay"                default="true"/>
  <arg name="replay_map"            default="replay_target_follow_1.bt"/>
  <arg name="master_name"            default="master"/>
  <arg name="slave_name"            default="slave"/>
  <arg name="enable_follower"       default="true" />
  <arg name="enable_target"       default="true" />
  <arg name="rviz_name"             default="rviz_config_follow.rviz" />
  <arg name="enable_rqtplot"        default="false" />
  
  <arg name="init_x" default="-8.0"/>
  <arg name="init_y" default="-4.0"/>
  <arg name="init_z" default=" 2.0"/>

  <arg name="end_x" default=" 8.0"/>
  <arg name="end_y" default=" 4.5"/>
  <arg name="end_z" default=" 2.0"/>

  <!-- Environment -->
  <arg name="world_x_min"           default="-10"/>
  <arg name="world_y_min"           default="-10"/>
  <arg name="world_z_min"           default="0.0"/>
  <arg name="world_x_max"           default="10"/>
  <arg name="world_y_max"           default="10"/>
  <arg name="world_z_max"           default="5"/>
  <arg name="world_resolution"      default="0.1"/>

  <!-- Obstacles -->
  <arg name="world_obs_num"         default="15"/>
  <arg name="world_r_min"           default="0.6"/>
  <arg name="world_r_max"           default="0.6"/>
  <arg name="world_h_min"           default="2.0"/>
  <arg name="world_h_max"           default="4.0"/>
  <arg name="world_margin"          default="0.5"/>
  
  <!-- InitTrajPlanner Parameters -->
  <arg name="ecbs_w"                default="1.3"/>
  <arg name="grid_xy_res"           default="0.5"/>
  <arg name="grid_z_res"            default="0.5"/>
  <arg name="grid_margin"           default="0.2"/>

  <!-- Corridor Parameters-->
  <arg name="box_xy_res"            default="0.1"/>
  <arg name="box_z_res"             default="0.1"/>

  <!-- RBPPlanner Parameters -->
  <arg name="plan_time_step"        default="1"/>
  <arg name="plan_time_step_slave"  default="1"/>
  <arg name="plan_downwash"         default="2.0"/>
  <arg name="plan_n"                default="5"/>
  <arg name="plan_phi"              default="3"/>  

  <arg name="plan_sequential"       default="false"/>
  <arg name="plan_batch_size"       default="4"/>
  <arg name="plan_N_b"              default="0"/>
  <arg name="plan_iteration"        default="1"/> 

<!-- Arguments End -->

<!-- Nodes Start -->
  <node pkg="octomap_server" type="octomap_server_node" name="octomap_server" unless="$(arg replay)">
    <remap from="cloud_in" to="/random_map_generator/all_map" />
    <param name="resolution"                 value="0.1" />
    <param name="frame_id"                   value="map" type="string"/>
    <param name="sensor_model/max_range"     value="20.0" />
    <param name="height_map"                 value="false"/>
    <param name="color/r"                    value="0.2" />
    <param name="color/g"                    value="0.2" />
    <param name="color/b"                    value="0.2" />
    <param name="color/a"                    value="0.2" />
  </node>

  <node pkg="octomap_server" type="octomap_server_node" name="octomap_server_replay" args="$(find swarm_planner)/worlds/$(arg replay_map)" if="$(arg replay)">
    <remap from="/octomap_point_cloud_centers" to="/random_map_generator/all_map" />
    <param name="resolution"                 value="0.1" />
    <param name="frame_id"                   value="map" type="string"/>
    <param name="sensor_model/max_range"     value="20.0" />
    <param name="height_map"                 value="false"/>
    <param name="color/r"                    value="0.2" />
    <param name="color/g"                    value="0.2" />
    <param name="color/b"                    value="0.2" />
    <param name="color/a"                    value="0.2" />
  </node>

  <node pkg="swarm_planner" type="random_map_generator" name="random_map_generator" unless="$(arg replay)">
    <param name="mission"                    value="$(find swarm_planner)/missions/$(arg mission)" />

    <param name="world/x_min"                value="$(arg world_x_min)" />
    <param name="world/y_min"                value="$(arg world_y_min)" />
    <param name="world/z_min"                value="$(arg world_z_min)" />
    <param name="world/x_max"                value="$(arg world_x_max)" />
    <param name="world/y_max"                value="$(arg world_y_max)" />
    <param name="world/z_max"                value="$(arg world_z_max)" />
    <param name="world/resolution"           value="$(arg world_resolution)" />
    
    <param name="world/obs_num"              value="$(arg world_obs_num)" />
    <param name="world/r_min"                value="$(arg world_r_min)" />
    <param name="world/r_max"                value="$(arg world_r_max)" />
    <param name="world/h_min"                value="$(arg world_h_min)" />
    <param name="world/h_max"                value="$(arg world_h_max)" />
    <param name="world/margin"               value="$(arg world_margin)" />
  </node>

  <node pkg="swarm_planner" type="swarm_traj_planner_rbp" name="swarm_traj_planner_rbp" output="screen" if="$(arg enable_target)" launch-prefix="xterm -e gdb --args">
    <param name="log"                        value="$(arg log)" />
    <param name="mission"                    value="$(find swarm_planner)/missions/$(arg mission)" />
    <param name="mav_name"                   value="$(arg master_name)" />
    <param name="world/x_min"                value="$(arg world_x_min)" />
    <param name="world/y_min"                value="$(arg world_y_min)" />
    <param name="world/z_min"                value="$(arg world_z_min)" />
    <param name="world/x_max"                value="$(arg world_x_max)" />
    <param name="world/y_max"                value="$(arg world_y_max)" />
    <param name="world/z_max"                value="$(arg world_z_max)" />

    <param name="ecbs/w"                     value="$(arg ecbs_w)" />
    <param name="grid/xy_res"                value="$(arg grid_xy_res)" />
    <param name="grid/z_res"                 value="$(arg grid_z_res)" />
    <param name="grid/margin"                value="$(arg grid_margin)" />

    <param name="box/xy_res"                 value="$(arg box_xy_res)" />
    <param name="box/z_res"                  value="$(arg box_z_res)" />

    <param name="plan/time_step"             value="$(arg plan_time_step)" />
    <param name="plan/downwash"              value="$(arg plan_downwash)" />
    <param name="plan/n"                     value="$(arg plan_n)" />
    <param name="plan/phi"                   value="$(arg plan_phi)" />

    <param name="plan/sequential"            value="$(arg plan_sequential)" />
    <param name="plan/batch_size"            value="$(arg plan_batch_size)" />
    <param name="plan/N_b"                   value="$(arg plan_N_b)" />
    <param name="plan/iteration"             value="$(arg plan_iteration)" />
  </node>

  <node pkg="swarm_planner" type="MPF_Follower" name="MPF_Follower" output="screen" launch-prefix="xterm -e gdb --args">
      <remap from="~waypoints"       to="/waypoint_generator/waypoints"/>
      <remap from="~odometry"        to="/odom/fake_odom"/>
      <remap from="~PointCloud"      to="/map_server/local_map" />
      <remap from="~position_cmd"    to="/position_cmd"  />
    
      <param name="log"                        value="$(arg log)" />
      <param name="mission"                    value="$(find swarm_planner)/missions/$(arg mission_follwer)" />
      <param name="mav_name"                   value="$(arg slave_name)" />
      <param name="world/x_min"                value="$(arg world_x_min)" />
      <param name="world/y_min"                value="$(arg world_y_min)" />
      <param name="world/z_min"                value="$(arg world_z_min)" />
      <param name="world/x_max"                value="$(arg world_x_max)" />
      <param name="world/y_max"                value="$(arg world_y_max)" />
      <param name="world/z_max"                value="$(arg world_z_max)" />

      <param name="ecbs/w"                     value="$(arg ecbs_w)" />
      <param name="grid/xy_res"                value="$(arg grid_xy_res)" />
      <param name="grid/z_res"                 value="$(arg grid_z_res)" />
      <param name="grid/margin"                value="$(arg grid_margin)" />

      <param name="box/xy_res"                 value="$(arg box_xy_res)" />
      <param name="box/z_res"                  value="$(arg box_z_res)" />

      <param name="plan/time_step"             value="$(arg plan_time_step_slave)" />
      <param name="plan/downwash"              value="$(arg plan_downwash)" />
      <param name="plan/n"                     value="$(arg plan_n)" />
      <param name="plan/phi"                   value="$(arg plan_phi)" />

      <param name="plan/sequential"            value="$(arg plan_sequential)" />
      <param name="plan/batch_size"            value="$(arg plan_batch_size)" />
      <param name="plan/N_b"                   value="$(arg plan_N_b)" />
      <param name="plan/iteration"             value="$(arg plan_iteration)" />

      <param name="planParam/plan_rate"        value="10.0"/>
      <param name="planParam/safety_margin"    value="0.5" />
      <param name="planParam/search_margin"    value="0.0" />
      <param name="planParam/max_radius"       value="5.0" />
      <param name="planParam/sensing_range"    value="12.0" />
      <param name="planParam/refine_portion"   value="0.7"  />
      <param name="planParam/sample_portion"   value="0.3" />
      <param name="planParam/goal_portion"     value="0.05" />
      <param name="planParam/path_find_limit"  value="0.1" />
      <param name="planParam/max_samples"      value="5000" />
      <param name="planParam/stop_horizon"     value="3.0"  />
      <param name="planParam/commitTime"       value="7.0"/>


      <param name="init_x"  value="$(arg init_x)"/>
      <param name="init_y"  value="$(arg init_y)"/>
      <param name="init_z"  value="$(arg init_z)"/>
      <param name="demoParam/target_x"     value="$(arg end_x)" />
      <param name="demoParam/target_y"     value="$(arg end_y)" />
      <param name="demoParam/target_z"     value="$(arg end_z)" />
      <param name="demoParam/goal_input"   value="false"  />
      <param name="demoParam/is_limit_vel" value="true"  />
      <param name="demoParam/is_limit_acc" value="false" />
      <param name="demoParam/is_print"     value="false" />

    </node>

  <node pkg="swarm_planner" type = "MPF_traj_server" output = "screen" name = "trajectory_server"  if="$(arg enable_follower)" >
          <remap  from = "~trajectory"        to = "/MPF_Follower/trajectory"/>
          <remap  from = "~odometry"          to = "/odom/fake_odom"/>
          <remap  from = "~position_command"  to = "/position_cmd"/>
          <param name="optimization/poly_order_min"           value="5"/> 
          <param name="optimization/poly_order_max"           value="5"/> 
  </node>
  <node pkg="swarm_planner" type = "odom_generator" output = "screen" name = "odom_generator"  if="$(arg enable_follower)">   
    <remap from = "~odometry"  to = "/odom/fake_odom"/>
    <remap from = "~command"   to = "/position_cmd"/>
    <param name="init_x"  value="$(arg init_x)"/>
    <param name="init_y"  value="$(arg init_y)"/>
    <param name="init_z"  value="$(arg init_z)"/>
  </node>

  <node pkg = "waypoint_generator" name = "waypoint_generator" type = "waypoint_generator" output = "screen">
          <remap from="~odom" to="/odom/fake_odom"/>        
          <remap from="~goal" to="/goal"/>
          <param name="waypoint_type" value="manual"/>    
  </node>
  <node pkg="odom_visualization" name="odom_visualization" type="odom_visualization" output="screen">
          <remap from="~cmd"  to="/position_cmd"/>
          <param name="color/a" value="0.8"/>    
          <param name="color/r" value="1.0"/>        
          <param name="color/g" value="0.0"/>        
          <param name="color/b" value="0.0"/>       
          <param name="covariance_scale" value="100.0"/>   
          <param name="robot_scale" value="1.0"/>       
          <param name="frame_id"    value="map"/>      
  </node>

  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find swarm_planner)/launch/rviz_config/$(arg rviz_name)" if="$(arg runsim)"/>
  <node pkg="rqt_plot" type="rqt_plot" name="position" args="/position_cmd/position/x:y:z" if="$(arg enable_rqtplot)"/>
  <node pkg="rqt_plot" type="rqt_plot" name="velocity" args="/position_cmd/velocity/x:y:z" if="$(arg enable_rqtplot)"/>
  <node pkg="rqt_plot" type="rqt_plot" name="acceleration" args="/position_cmd/acceleration/x:y:z" if="$(arg enable_rqtplot)"/>

<!-- Nodes End -->
</launch>
